File: programmingtheiot/cda/connection/CoapClientConnector.py
import logging
import socket
import traceback

from coapthon import defines
from coapthon.client.helperclient import HelperClient
from coapthon.messages.option import Option
from coapthon.utils import generate_random_token

import programmingtheiot.common.ConfigConst as ConfigConst
from programmingtheiot.common.ConfigUtil import ConfigUtil
from programmingtheiot.common.ResourceNameEnum import ResourceNameEnum
from programmingtheiot.common.IDataMessageListener import IDataMessageListener
from programmingtheiot.cda.connection.IRequestResponseClient import IRequestResponseClient
from programmingtheiot.data.DataUtil import DataUtil


class CoapClientConnector(IRequestResponseClient):
    """
    Implements a CoAP Client using the CoAPthon3 library.
    Supports GET, POST, PUT, DELETE, DISCOVERY, and OBSERVE.
    """

    def __init__(self, dataMsgListener: IDataMessageListener = None):
        self.config = ConfigUtil()
        self.dataMsgListener = dataMsgListener
        self.enableConfirmedMsgs = False
        self.coapClient = None
        self.observeRequests = {}
        self.includeDebugLogDetail = True

        # Load configuration
        self.host = self.config.getProperty(ConfigConst.COAP_GATEWAY_SERVICE,
                                            ConfigConst.HOST_KEY,
                                            ConfigConst.DEFAULT_HOST)
        self.port = self.config.getInteger(ConfigConst.COAP_GATEWAY_SERVICE,
                                           ConfigConst.PORT_KEY,
                                           ConfigConst.DEFAULT_COAP_PORT)

        try:
            tmpHost = socket.gethostbyname(self.host)
            if tmpHost:
                self.host = tmpHost
                self.uriPath = f"coap://{self.host}:{self.port}/"
                logging.info(f"CoAP client will connect to: {self.uriPath}")
                self._initClient()
            else:
                logging.error(f"Cannot resolve host: {self.host}")
        except socket.gaierror as e:
            logging.error(f"Failed to resolve host: {self.host}")
            raise e

    def _initClient(self):
        try:
            self.coapClient = HelperClient(server=(self.host, self.port))
            logging.info(f"CoAP Client created. Connected to {self.host}:{self.port}")
        except Exception as e:
            logging.error("Failed to initialize CoAP client.")
            traceback.print_exception(type(e), e, e.__traceback__)

    # --------------------------------------------------------
    # Utility and setup methods
    # --------------------------------------------------------
    def setDataMessageListener(self, listener: IDataMessageListener = None) -> bool:
        if listener:
            self.dataMsgListener = listener
            return True
        return False

    def _createResourcePath(self, resource: ResourceNameEnum = None, name: str = None):
        resourcePath = ""
        if resource:
            resourcePath += resource.value
        if name:
            if resourcePath:
                resourcePath += "/"
            resourcePath += name
        return resourcePath

    # --------------------------------------------------------
    # DISCOVERY
    # --------------------------------------------------------
    def sendDiscoveryRequest(self, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        logging.info("Discovering remote CoAP resources...")
        try:
            resourcePath = ".well-known/core"
            request = self.coapClient.mk_request(defines.Codes.GET, path=resourcePath)
            request.token = generate_random_token(2)
            self.coapClient.send_request(request, timeout=timeout, callback=self._onDiscoveryResponse)
            return True
        except Exception as e:
            logging.error(f"Discovery request failed: {e}")
            return False

    def _onDiscoveryResponse(self, response):
        if not response:
            logging.warning("DISCOVERY response invalid. Ignoring.")
            return
        logging.info(f"DISCOVERY response received: {response.payload}")

    # --------------------------------------------------------
    # GET
    # --------------------------------------------------------
    def sendGetRequest(self, resource: ResourceNameEnum = None, name: str = None,
                       enableCON: bool = False, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        try:
            resourcePath = self._createResourcePath(resource, name)
            logging.info(f"Issuing GET request to path: {resourcePath}")
            request = self.coapClient.mk_request(defines.Codes.GET, path=resourcePath)
            request.token = generate_random_token(2)
            if not enableCON:
                request.type = defines.Types["NON"]
            response = self.coapClient.send_request(request, timeout=timeout)
            self._onGetResponse(response, resourcePath)
            return True
        except Exception as e:
            logging.error(f"GET request failed: {e}")
            return False

    def _onGetResponse(self, response, resourcePath: str = None):
        if not response:
            logging.warning("GET response invalid. Ignoring.")
            return
        logging.info("GET response received.")
        jsonData = response.payload
        logging.info(f"Response data: {jsonData}")
        try:
            dataType = resourcePath.split('/')[-1] if resourcePath else ""
            if dataType and ConfigConst.ACTUATOR_CMD in dataType:
                ad = DataUtil().jsonToActuatorData(jsonData)
                if self.dataMsgListener:
                    self.dataMsgListener.handleActuatorCommandMessage(ad)
        except Exception as e:
            logging.warning(f"Failed to decode GET payload: {e}")

    # --------------------------------------------------------
    # PUT
    # --------------------------------------------------------
    def sendPutRequest(self, resource: ResourceNameEnum = None, name: str = None,
                       enableCON: bool = False, payload: str = None,
                       timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        try:
            resourcePath = self._createResourcePath(resource, name)
            logging.info(f"Issuing PUT with path: {resourcePath}")
            request = self.coapClient.mk_request(defines.Codes.PUT, path=resourcePath)
            request.token = generate_random_token(2)
            request.payload = payload
            if not enableCON:
                request.type = defines.Types["NON"]
            self.coapClient.send_request(request, callback=self._onPutResponse, timeout=timeout)
            return True
        except Exception as e:
            logging.error(f"PUT request failed: {e}")
            return False

    def _onPutResponse(self, response):
        if not response:
            logging.warning("PUT response invalid. Ignoring.")
            return
        logging.info(f"PUT response received: {response.payload}")

    # --------------------------------------------------------
    # POST
    # --------------------------------------------------------
    def sendPostRequest(self, resource: ResourceNameEnum = None, name: str = None,
                        enableCON: bool = False, payload: str = None,
                        timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        try:
            resourcePath = self._createResourcePath(resource, name)
            logging.info(f"Issuing POST with path: {resourcePath}")
            request = self.coapClient.mk_request(defines.Codes.POST, path=resourcePath)
            request.token = generate_random_token(2)
            request.payload = payload
            if not enableCON:
                request.type = defines.Types["NON"]
            self.coapClient.send_request(request, callback=self._onPostResponse, timeout=timeout)
            return True
        except Exception as e:
            logging.error(f"POST request failed: {e}")
            return False

    def _onPostResponse(self, response):
        if not response:
            logging.warning("POST response invalid. Ignoring.")
            return
        logging.info(f"POST response received: {response.payload}")

    # --------------------------------------------------------
    # DELETE
    # --------------------------------------------------------
    def sendDeleteRequest(self, resource: ResourceNameEnum = None, name: str = None,
                          enableCON: bool = False, timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        try:
            resourcePath = self._createResourcePath(resource, name)
            logging.info(f"Issuing DELETE with path: {resourcePath}")
            request = self.coapClient.mk_request(defines.Codes.DELETE, path=resourcePath)
            request.token = generate_random_token(2)
            if not enableCON:
                request.type = defines.Types["NON"]
            self.coapClient.send_request(request, callback=self._onDeleteResponse, timeout=timeout)
            return True
        except Exception as e:
            logging.error(f"DELETE request failed: {e}")
            return False

    def _onDeleteResponse(self, response):
        if not response:
            logging.warning("DELETE response invalid. Ignoring.")
            return
        logging.info(f"DELETE response received: {response.payload}")

    # --------------------------------------------------------
    # OBSERVE
    # --------------------------------------------------------
    def startObserver(self, resource: ResourceNameEnum = None, name: str = None,
                      ttl: int = IRequestResponseClient.DEFAULT_TTL) -> bool:
        if resource in self.observeRequests:
            logging.warning(f"Already observing resource {resource}. Ignoring.")
            return False

        resourcePath = self._createResourcePath(resource, name)
        handler = HandleActuatorEvent(listener=self.dataMsgListener,
                                      resource=resource,
                                      requests=self.observeRequests)
        try:
            self.coapClient.observe(path=resourcePath,
                                    callback=handler.handleActuatorResponse)
            self.observeRequests[resource] = handler
            logging.info(f"Started observing resource: {resourcePath}")
            return True
        except Exception as e:
            logging.warning(f"Failed to start observer for {resourcePath}: {e}")
            return False

    def stopObserver(self, resource: ResourceNameEnum = None, name: str = None,
                     timeout: int = IRequestResponseClient.DEFAULT_TIMEOUT) -> bool:
        if not resource or resource not in self.observeRequests:
            logging.warning(f"Resource {resource} not being observed.")
            return False
        handler = self.observeRequests.pop(resource, None)
        if handler:
            logging.info(f"Stopped observing resource {resource}.")
            return True
        return False


class HandleActuatorEvent:
    """
    Internal class for handling ActuatorData observation events.
    """

    def __init__(self, listener: IDataMessageListener = None,
                 resource: ResourceNameEnum = ResourceNameEnum.CDA_ACTUATOR_CMD_RESOURCE,
                 requests=None):
        self.listener = listener
        self.resource = resource
        self.observeRequests = requests or {}

    def handleActuatorResponse(self, response):
        if response:
            jsonData = response.payload
            logging.info(f"Received actuator response for {self.resource}: {jsonData}")
            if self.listener:
                try:
                    data = DataUtil().jsonToActuatorData(jsonData)
                    self.listener.handleActuatorCommandMessage(data)
                except Exception as e:
                    logging.warning(f"Failed to decode actuator observe data: {e}")

File: src/main/java/programmingtheiot/gda/connection/CoapClientConnector.java
package programmingtheiot.gda.connection;

import java.util.logging.Logger;

import org.eclipse.californium.core.CoapClient;
import org.eclipse.californium.core.CoapHandler;
import org.eclipse.californium.core.CoapResponse;
import org.eclipse.californium.core.coap.CoAP;
import org.eclipse.californium.core.coap.Request;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ConfigUtil;

/**
 * CoAP client implementation for the GDA.
 * Fully compatible with Lab 10 performance tests.
 */
public class CoapClientConnector implements IRequestResponseClient
{
    private static final Logger _Logger =
        Logger.getLogger(CoapClientConnector.class.getName());

    private CoapClient clientConn = null;
    private IDataMessageListener dataMsgListener = null;

    // CoAP endpoint configuration
    private String host = "localhost";
    private int port = 5683;
    private boolean enableCON = true;      // Used for Lab 10 tests

    public CoapClientConnector() {
        super();
        initClient();
    }

    private void initClient() {
        ConfigUtil cfg = ConfigUtil.getInstance();

        try { this.host = cfg.getProperty(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.HOST_KEY, "localhost"); }
        catch (Exception e) { this.host = "localhost"; }

        try { this.port = cfg.getInteger(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.PORT_KEY, 5683); }
        catch (Exception e) { this.port = 5683; }

        try { this.enableCON = cfg.getBoolean(ConfigConst.COAP_GATEWAY_SERVICE, ConfigConst.ENABLE_CON_MSGS_KEY); }
        catch (Exception e) { this.enableCON = true; }

        String baseUri = "coap://" + this.host + ":" + this.port;
        this.clientConn = new CoapClient(baseUri);
    }

    private String buildUri(ResourceNameEnum resource) {
        return "coap://" + this.host + ":" + this.port + "/" + resource.getResourceName();
    }

    // ======================================================
    // Interface: IRequestResponseClient
    // ======================================================

    @Override
    public void setEndpointPath(ResourceNameEnum resource) {
        if (resource != null) {
            this.clientConn.setURI(buildUri(resource));
        }
    }

    @Override
    public void clearEndpointPath() {
        this.clientConn.setURI("coap://" + this.host + ":" + this.port);
    }

    @Override
    public boolean setDataMessageListener(IDataMessageListener listener) {
        if (listener != null) {
            this.dataMsgListener = listener;
            return true;
        }
        return false;
    }

    @Override
    public boolean sendGetRequest(ResourceNameEnum resource, String name, boolean useCON, int timeout) {
        try {
            setEndpointPath(resource);

            Request req = new Request(CoAP.Code.GET, (useCON ? CoAP.Type.CON : CoAP.Type.NON));
            CoapResponse resp = this.clientConn.advanced(req);

            if (resp != null && this.dataMsgListener != null) {
                this.dataMsgListener.handleIncomingMessage(resource, resp.getResponseText());
            }

            return resp != null;
        }
        catch (Exception e) {
            _Logger.warning("GET request failed: " + e.getMessage());
            return false;
        }
        finally {
            clearEndpointPath();
        }
    }

    @Override
    public boolean sendPostRequest(ResourceNameEnum resource, String name, boolean useCON,
                                   String payload, int timeout) {
        try {
            setEndpointPath(resource);

            Request req = new Request(CoAP.Code.POST, (useCON ? CoAP.Type.CON : CoAP.Type.NON));
            req.setPayload(payload);

            CoapResponse resp = this.clientConn.advanced(req);

            if (resp != null && this.dataMsgListener != null) {
                this.dataMsgListener.handleIncomingMessage(resource, resp.getResponseText());
            }

            return resp != null;
        }
        catch (Exception e) {
            _Logger.warning("POST request failed: " + e.getMessage());
            return false;
        }
        finally {
            clearEndpointPath();
        }
    }

    @Override
    public boolean sendPutRequest(ResourceNameEnum resource, String name, boolean useCON,
                                  String payload, int timeout) {
        try {
            setEndpointPath(resource);

            Request req = new Request(CoAP.Code.PUT, (useCON ? CoAP.Type.CON : CoAP.Type.NON));
            req.setPayload(payload);

            CoapResponse resp = this.clientConn.advanced(req);

            if (resp != null && this.dataMsgListener != null) {
                this.dataMsgListener.handleIncomingMessage(resource, resp.getResponseText());
            }

            return resp != null;
        }
        catch (Exception e) {
            _Logger.warning("PUT request failed: " + e.getMessage());
            return false;
        }
        finally {
            clearEndpointPath();
        }
    }

    @Override
    public boolean sendDeleteRequest(ResourceNameEnum resource, String name, boolean useCON, int timeout) {
        try {
            setEndpointPath(resource);

            Request req = new Request(CoAP.Code.DELETE, (useCON ? CoAP.Type.CON : CoAP.Type.NON));
            CoapResponse resp = this.clientConn.advanced(req);

            if (resp != null && this.dataMsgListener != null) {
                this.dataMsgListener.handleIncomingMessage(resource, resp.getResponseText());
            }

            return resp != null;
        }
        catch (Exception e) {
            _Logger.warning("DELETE request failed: " + e.getMessage());
            return false;
        }
        finally {
            clearEndpointPath();
        }
    }

    @Override
    public boolean sendDiscoveryRequest(int timeout) {
        try {
            this.clientConn.setURI("coap://" + this.host + ":" + this.port + "/.well-known/core");
            CoapResponse resp = this.clientConn.get();

            if (resp != null) {
                _Logger.info("Discovery response: " + resp.getResponseText());
                return true;
            }
        }
        catch (Exception e) {
            _Logger.warning("Discovery request failed: " + e.getMessage());
        }
        finally {
            clearEndpointPath();
        }

        return false;
    }

    @Override
    public boolean startObserver(ResourceNameEnum resource, String name, int ttl) {
        try {
            setEndpointPath(resource);

            this.clientConn.observe(new CoapHandler() {
                @Override
                public void onLoad(CoapResponse resp) {
                    if (resp != null && dataMsgListener != null) {
                        dataMsgListener.handleIncomingMessage(resource, resp.getResponseText());
                    }
                }

                @Override
                public void onError() {
                    _Logger.warning("Observer failed for resource: " + resource.getResourceName());
                }
            });

            return true;
        }
        catch (Exception e) {
            _Logger.warning("Observer setup failed: " + e.getMessage());
            return false;
        }
    }

    @Override
    public boolean stopObserver(ResourceNameEnum resource, String name, int timeout) {
        try {
            this.clientConn.shutdown();
            return true;
        }
        catch (Exception e) {
            _Logger.warning("Stop observer failed: " + e.getMessage());
            return false;
        }
    }

    // ---------------------------------------------------------
    // Helper: connect / disconnect client
    // ---------------------------------------------------------
    public void connectClient() {
        if (this.clientConn == null) initClient();
    }

    public void disconnectClient() {
        if (this.clientConn != null) {
            this.clientConn.shutdown();
            this.clientConn = null;
        }
    }

    // Optional getters/setters
    public void setUseConMessages(boolean useCon) { this.enableCON = useCon; }
    public boolean getUseConMessages() { return this.enableCON; }
    public void setHost(String host) { this.host = host; }
    public void setPort(int port) { this.port = port; }
}
